### ISSUE TITLE
Add introspection for storage collection frontiers

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#13489]
### Feature request

For performance testing it would be useful to have a way to know, to which times storage collections have already advanced. This would allow us to answer questions like "Is this peek slow because it is waiting on the compute dataflow to catch up, or is it waiting for the data to be available in storage?".

We already track frontiers of compute collections as part of the `computed` introspection (see `mz_materialization_frontiers`). The proposal is to add the same introspection capability for storage collections as well.

**Comment by teskje** [comment]
We need to think about what is the best way to implement this. Where should the introspection data be generated? `storaged` would make sense for collections that are written by `storaged`, but not all of them are. For example:
* tables are written by the storage controller
* recorded views are written by `computed`

The storage controller eventually sees all frontier advancements, but it doesn't get them from persist directly and relies on other components (`storaged`s or the compute controller) to report them. This means it doesn't know the actual times of frontier advancements. There are discussions about making the storage controller watch for frontier advancements on the persist shards directly, which might solve this issue (not sure!).

Are there other viable options?

**Comment by teskje** [comment]
Another concern is that if we want to actually compare timings of introspection events, they need to be observed on the same system. If we'd monitor storage collection frontier advancements in `storaged` or the storage controller, we couldn't compare their times to compute collection frontier advancements collected in `computed`, because the different systems have different clocks that might be misaligned. We could probably compare on coarse timescales (minutes), but probably not on the microsecond-timescales we need for debugging peek times.

This means that `computed` is really the only component that can implement this kind of storage frontier tracking that is useful for debugging peeks.

@antiguru pointed out that the same issue exists between the different `computed` instances in a multi-process timely cluster. I think we can solve this by tracking storage frontiers separately on each timely process, like we already do for compute frontiers (see `mz_worker_materialization_frontiers`, which contains a `worker` column).

**Comment by antiguru** [comment]
> @antiguru pointed out that the same issue exists between the different computed instances in a multi-process timely cluster. I think we can solve this by tracking storage frontiers separately on each timely process, like we already do for compute frontiers (see mz_worker_materialization_frontiers, which contains a worker column).

For Timely, it's a slightly different problem because the loggers use the time after establishing the connections as their starting point. This might or might not be closely synchronized, I can't say.

**Comment by frankmcsherry** [comment]
> Another concern is that if we want to actually compare timings of introspection events, they need to be observed on the same system.

I think one thing `computed` will want to do independent of `storaged` logging is report the observed frontier at each source *instantiation*. So, each time we build up a new persist source, its frontier is of interest to us, and we should log that. This is the actual thing that indicates to the `computed` when data are available. The gap between `storaged` receiving data and it being available through `persist` is also interesting, but at least the source instance frontier is within `computed`'s power to observe.

**Comment by vmarcos** [comment]
After some discussion, we have observed that it is important to provide two important kinds of introspection data related to the lag between source and dataflow frontiers: (a) At a given wall-clock time point, the current source and the current dataflow frontiers in Materialize logical times; (b) The histogram of differences between the two wall-clock time points at which COMPUTE observes the advancements of the frontiers of source instantiations through persist and the later advancements of the corresponding dataflow frontiers.

It is proposed that point (a) be handled by this issue, since it requires information on the current source frontier to be provided by the STORAGE layer. 

Point (b) has been moved to a separate issue that can be handled entirely within COMPUTE (#13885).

**Comment by frankmcsherry** [comment]
I'm also up for (c) which is (a) but done at COMPUTE. I'm not sure we shook out the (b) vs (c) trade-offs. But, for sure at least one of (b) or (c) is important for COMPUTE to be able to disconfirm that it is holding the ball on some query.
